opendevin/controller/agent_controller.py

```diff
@@ -195,45 +206,88 @@ def get_agent_state(self):
         self.delegate = AgentController(
             sid=self.id + '-delegate',
             agent=agent,
             event_stream=self.event_stream,
-            max_iterations=self.max_iterations,
+            max_iterations=self.state.max_iterations,
             max_chars=self.max_chars,
-            inputs=action.inputs,
+            initial_state=state,
+            is_delegate=True,
         )
+        await self.delegate.set_agent_state_to(AgentState.RUNNING)
 
     async def _step(self):
+        logger.debug(f'[Agent Controller {self.id}] Entering step method')
         if self.get_agent_state() != AgentState.RUNNING:
-            logger.debug('waiting for agent to run...')
+            logger.info(f'[Agent Controller {self.id}] waiting for agent to run...')
             await asyncio.sleep(1)
             return
 
         if self._pending_action:
-            logger.debug('waiting for pending action: ' + str(self._pending_action))
+            logger.info(
+                f'[Agent Controller {self.id}] waiting for pending action: {self._pending_action}'
+            )
             await asyncio.sleep(1)
             return
 
-        logger.info(f'STEP {self.state.iteration}', extra={'msg_type': 'STEP'})
-        if self.state.iteration >= self.max_iterations:
-            await self.report_error('Agent reached maximum number of iterations')
-            await self.set_agent_state_to(AgentState.ERROR)
-            return
-
         if self.delegate is not None:
-            delegate_done = await self.delegate._step()
+            logger.debug(f'[Agent Controller {self.id}] Delegate not none, awaiting...')
+            assert self.delegate != self
+            await self.delegate._step()
+            logger.debug(f'[Agent Controller {self.id}] Delegate step done')
+            assert self.delegate is not None
+            delegate_state = self.delegate.get_agent_state()
+            if delegate_state == AgentState.ERROR:
+                # close the delegate upon error
+                await self.delegate.close()
+                await self.report_error('Delegator agent encounters an error')
+                # propagate error state until an agent or user can handle it
+                await self.set_agent_state_to(AgentState.ERROR)
+                return
+            delegate_done = delegate_state == AgentState.FINISHED
             if delegate_done:
+                logger.info(
+                    f'[Agent Controller {self.id}] Delegate agent has finished execution'
+                )
+                # retrieve delegate result
                 outputs = self.delegate.state.outputs if self.delegate.state else {}
-                obs: Observation = AgentDelegateObservation(content='', outputs=outputs)
-                await self.event_stream.add_event(obs, EventSource.AGENT)
+
+                # update current controller's state
+                self.state.history += self.delegate.state.history
```

Comment:
Please help me understand this a little, now that the delegate is done, does the parent need to send the child's historical messages to the LLM going forward?

Or is there another purpose for updating itself with its history?

Reply:
Originally suggested by @rbren here: https://github.com/OpenDevin/OpenDevin/pull/1910#discussion_r1612023065

I think the rationale is, the parent might be interested in learning what the child has done, to drive the task forward. It is debatable, though, that the parent may only need to know the result/summary of its child. But should parent trust child's conclusion? What if the child is executed with a weaker LLM (not possible at the moment but maybe in the future to lower down the cost - parent agent using more powerful LLM while delegates use small modes). Would a memory condenser be interesting? These open questions are hard to answer without benchmarking...

Reply:
Yeah I've gone back and forth on this.

I think keeping the _parent's_ memory out of the child makes a lot of sense. I could also see a case for keeping the child's memory out of the parent.

I forget what the problem was that led me to that original comment, but it definitely broke something I was working on...

Reply:
I say we get this in either way, and we can take another look at history management soon

Reply:
Yeah let's keep it this way.

Reply:
I changed my mind when I was working on #1945. The parent agent would get confused when it sees the child agent's history because the parent doesn't know those commands used by child agent.

Reply:
So now, the parent agent only learns about the result provided by the child agent.

Reply:
It might be pure luck, but after removing child history, I ran regenerate and ManagerAgent was able to finish the test tasks in fewer turns.

Reply:
Interesting! Thanks for the details. Hmm, food for thought... What LLM did you use in the tests?

Reply:
I was using gpt-4o (and gpt-4-turbo, sometimes).